content/learning-paths/cross-platform/integer-vs-floats/integer-float-conversions.md

```diff
@@ -0,0 +1,196 @@
+|            int32_t   |             |             |   int64_t   |    int128_t*  |
+|            int64_t   |             |             |             |    int128_t*  |
+|            fp16**    |             |    float    |   double    |  long double* |
+|            bf16**    |             |    float    |   double    |  long double* |
+|            float     |             |             |   double    |  long double* |
+|            double    |             |             |             |  long double* |
+
+(`*`) depends on the actual compiler support
+(`**`) depends on hardware support
+
+Similar promotions take place for unsigned integers.
+
+### Demotions
+
+Respectively, if `A` is `float` and `B` is `double` but `C` is `float` then we have a demotion.
+This is actually a very risky conversion, as bits are lost and it is discouraged and should be avoided unless in specific circumstances when you know exactly what you are doing.
+Unfortunately, this is where the programming language matters. In some cases, C++ will catch such a demotion (called a *narrowing conversion* in C++) and will issue a relevant warning,
+C does not allow for such a warning. The C compiler will not issue a warning and it's easy for a conversion bug to creep in your code, and sometimes these can be very hard to detect bugs.
+Even on C++ there is a catch however, the compiler will only issue such a warning when using bracket initialization, not assignment between values. You will have a chance to see this in detail in the next section.
+
+Here is a similar list for possible demotions:
+
+| Data type (stdint.h)  | Demoted to                                             |
+| -------------------- | ----------- | ----------- | ----------- | ------------- |
+|            int16_t   |   int8_t    |             |             |               |
+|            int32_t   |   int8_t    |   int16_t   |             |               |
+|            int64_t   |   int8_t    |   int16_t   |   int32_t   |               |
+|            fp16**    |             |    float    |   double    |  long double* |
+|            bf16**    |             |    float    |   double    |  long double* |
+|            float     |             |             |   double    |  long double* |
+|            double    |             |             |             |  long double* |
+
+Again unsigned integers are demoted to similar types.
+
+(`*`) depends on the actual compiler support
+(`**`) depends on hardware support
+
+### Type conversions
+
+Some might argue that conversion of an `int16_t` to `float` or `double` is a promotion, but it's really not. A promotion or demotion can be achieved in the CPU just by copying and either narrowing or widening the elements between registers, no conversion instruction involved. However when an integer is converted to a floating point number, or vice-versa, there is **always** a conversion instruction involved. And it's almost always more costly than a mere copy.
```

Comment:
Not sure I understand what this is trying to say.  promotion always have an instruction.
for integer it needs to zero or sign extend if promoting, demotion doesn't need instructions yes.

for float you always do since you need to preserve e.g. `NaN` and `Inf` and rounding has to be performed https://godbolt.org/z/4r4qqM1c3

Reply:
I was trying to say that propotion might involve an instruction, but essentially promotion between eg unsigned integers could be achieved without using an sign-extend instruction. By use of a shuffle-type instruction or a memory copy. Even sign-extend can be done this way. Converting integers to/from floating-point always involves a special instruction as the number representation changes. Anyway, I understand your point, I will rephrase it better.

Reply:
I see, I think we'd want to be careful there though, since compilers recognize sequences like zero extension and generate efficient sequences, for instance https://godbolt.org/z/Erhnnnfxb

Making the scalar code more complicated is likely to make us mis them.  So If you do want to cover them I think we should do it in the form of explaining the compiler output, so that users know their naive inputs are handled optimally.

Reply:
I have expanded on promotion/demotion based on your example.